lib/player.js

```diff
@@ -2093,21 +2093,49 @@ shaka.Player = class extends shaka.util.FakeEventTarget {
 
     const LanguageUtils = shaka.util.LanguageUtils;
     const preferredLanguage = LanguageUtils.normalize(preferredAudioLanguage);
-    const audioLanguages = this.getAudioLanguages();
-    for (const language of audioLanguages) {
+    const audioLanguagesAndRoles = this.getAudioLanguagesAndRoles();
+    const matchingByLanguage = [];
+    for (const languageAndRole of audioLanguagesAndRoles) {
+      const language = languageAndRole.language;
       if (LanguageUtils.normalize(language) == preferredLanguage) {
-        this.selectAudioLanguage(preferredLanguage);
-        break;
+        matchingByLanguage.push(languageAndRole);
       }
     }
+
+    // If there is no language match, the browser preference is left.
+    if (!matchingByLanguage.length) {
+      return;
+    }
+
+    const preferredVariantRole = this.config_.preferredVariantRole;
+
+    // If the user has not selected a role preference, the first match is
+    // selected.
+    if (preferredVariantRole == '') {
+      const languageAndRole = matchingByLanguage[0];
+      this.selectAudioLanguage(languageAndRole.language, languageAndRole.role);
+      return;
+    }
+
+    const matchingByLanguageAndRole = matchingByLanguage.find(
+        (languageAndRole) => languageAndRole.role == preferredVariantRole);
+    if (matchingByLanguageAndRole) {
+      this.selectAudioLanguage(matchingByLanguageAndRole.language,
+          matchingByLanguageAndRole.role);
+      return;
+    }
+
+    // If there is no role match, the first match is selected.
+    const languageAndRole = matchingByLanguage[0];
+    this.selectAudioLanguage(languageAndRole.language, languageAndRole.role);
   }
 
   /**
    * This method setup the preferred text using src=.
    *
    * @private
    */
-  setupPreferredTextLanguageOnSrc_() {
+  setupPreferredTextOnSrc_() {
```

Comment:
It seems that almost all of this logic is already encapsulated in selectTextLanguage(language, role).  It already filters by language, then looks for a matching role.  The only difference is the fallback when the preferred role isn't available.

Perhaps we could do away with this whole thing and instead call select twice:

```js
// If this second call can't find a match, then the first call gives
// us a fallback for a matching language with any role.
// If both calls fail, we use what the browser pre-selected.
this.selectTextLanguage(preferredTextLanguage);
this.selectTextLanguage(preferredTextLanguage, preferredTextRole);
```

This should work for audio, too.

Reply:
You're right, I'm going to change it now.